### DISCUSSION TITLE
Can I manually paint hard lines between textures?

### DISCUSSION THREAD

**Discussion by pedroyzkrak** [TokisanGames/Terrain3D#580] [Questions] [answered]
Due to the blurriness I've obtained while attempting to manually painting textures:

![image](https://github.com/user-attachments/assets/108f647b-fa8d-45e5-950b-95585c5899b5)

I've followed this discussion here: https://github.com/TokisanGames/Terrain3D/discussions/350?u=http://github.com/TokisanGames/Terrain3D/discussions/categories/questions

and disabled texture filtering, however I think that's not exactly what I'm looking for. 
My question is: is it possible to obtain this result (autoshader):

![image](https://github.com/user-attachments/assets/ccbef2ec-4ef1-454f-84e3-b6dcab9f52f0)

with manual texture painting/brushing? As expected by disabling texture filtering I get this:

![image](https://github.com/user-attachments/assets/e76d521e-2d15-46dc-b215-adfe206d4ad3)

Any hints or recommendations are welcome!

PS - for extra info my uv scales are really low (< .1) and my textures are 1k res and tried both linear and nearest texture filtering (on both texture import and Terrain3D material settings)

**Comment by TokisanGames** [comment, marked as answer]
This has nothing to do with texture filtering, which was explained on the other ticket and demonstrated here https://youtu.be/YtiAI2F6Xkk?t=537.

This is a duplicate of #562.

> My question is: is it possible to obtain this result (autoshader):

It is not possible to manually paint hard lines with the current shader.

As explained here, this is a vertex painter, not a pixel painter.
https://terrain3d.readthedocs.io/en/stable/docs/introduction.html#vertex-painting

The shader is setup to bilnearly interpolate among adjacent vertices for a blended result, which works well with natural looking textures.

The autoshader calculates based on slope, and you have removed the blend to get a sharp line.

Your options are:
* Dispense with manual shading and customize the shader to expand the autoshader to draw textures with hard lines wherever you want them. Difficult to do without some definite value you can build on. The autoshader textures based on slope, which is definitive. Height is another common one people use. There are other options, anything you can think of, eg within a radius around certain XZ coordinates.

* Customize the shader to change the way textures are interpolated between vertices. Instead of doing a bilerp (weighted average) shown in the other ticket, you would need to calculate a bezier curve between all vertices. It's a theroetical idea. I don't know how I would implement it.

* Decrease vertex spacing to the smallest. Perhaps this will give enough resolution. Only do this on a small terrain. 
 
* Use another terrain system or a terrain mesh where you can pixel paint. The terrain texture map will need to be much higher resolution than the vertices of the mesh. We maintain 1px = 1vertex. You probably need 10:1 or higher.

**Reply by pedroyzkrak** [reply]
Thanks so much for your recommendations :)
